src/Transfer.js

```diff
@@ -0,0 +1,463 @@
+     */
+    const [highlightedSelected, setHighlightedSelected] = useState([])
+    const toggleHighlightedSelected = ({ option, metaKeyPressed }) => {
+        if (disabled) return
+
+        setHighlightedOptions([])
+
+        if (
+            highlightedSelected.length > 1 &&
+            !metaKeyPressed &&
+            findOption(highlightedSelected, option)
+        ) {
+            setHighlightedSelected([option])
+        } else {
+            setHighlightedSelected(
+                toggleOption(highlightedSelected, option).slice(
+                    -1 * (metaKeyPressed ? maxSelections : 1)
+                )
+            )
+        }
+    }
+
+    /**
+     * Extract the selected options. This way custom options are supported
+     * without having to provide a component via the props
+     *
+     * Children are sorted by the order given in the "selected" array.
+     * This is done in order to cover the "append newly selected items
+     * at the end" feature/behavior.
+     */
+    const selectedElements = Children.toArray(allOptions)
+        .map(child => {
+            const selectedOptions = selected
+            const isSelected = !!findOption(selectedOptions, child.props)
+            const hasError =
+                isSelected && !!findOption(errorOptions, child.props)
+            const isHighlighted = !!findOption(highlightedSelected, child.props)
+
+            if (!isSelected) return null
+
+            return React.cloneElement(child, {
+                error: hasError,
+                highlighted: isHighlighted,
+                onClick: (option, event) => {
+                    const metaKeyPressed =
+                        event.altKey ||
+                        event.shiftKey ||
+                        event.ctrlKey ||
+                        event.metaKey
+
+                    toggleHighlightedSelected({ option, metaKeyPressed })
+                },
+            })
+        })
+        // We can ONLY do this because the children have keys
+        .filter(child => !!child)
+
+    selectedElements.sort((left, right) => {
+        const leftIndex = findOptionIndex(selected, left.props)
+        const rightIndex = findOptionIndex(selected, right.props)
+
+        if (leftIndex < rightIndex) return -1
+        return 1
+    })
+
+    /**
+     * Filtered options, always same as "availableOptions" when
+     * "enableFilter" is false
+     */
+    const filteredOptions = (() => {
+        const options = getOptionsFromChildren(availableOptions)
+
+        const filtered =
+            enableFilter && filter ? filterOptions(options, filter) : options
+
+        return Children.map(availableOptions, child => {
+            if (!enableFilter || filter === '') return child
+            if (!filtered.find(option => isOption(option, child.props)))
+                return null
+
+            return child
+        })
+    })()
+
+    return (
```

Comment:
There is a lot going on before we can get to return some JSX. I wonder whether all of this needs to be done on every render, and/or whether we could extract this into some dedicated helper functions....

Reply:
I've changed a lot in the `src/Transfer.js`. Please have a look

Reply:
Nice 👍